### ISSUE TITLE
[SPIRV] Incorrect Vulkan Result on Mobile GPU

### ISSUE THREAD

**Issue by tqchen** [apache/incubator-tvm#2355]
There have been several incidences reported about incorrect results when deploying to MobileGPU via vulkan. We will need to look into this. 

More specifically, it would be really useful to produce a minimum reduced example from the end to end ones, to use debug_runtime to get the output of each intermediate step and compare them with the CPU or OpenCL version. Then we can look into which specific setup causes the problem.

We would love to have volunteers from the community to look into this issue.

Related threads:

- https://discuss.tvm.ai/t/vulkan-deploy-issue-with-latest-code/604/7
- https://discuss.tvm.ai/t/huawei-p20-pro-vulkan-output-is-not-proper/702

cc @eqy 

Possible steps to debug the issue:

- Get an end to end example that produces the wrong result
- Use debug runtime to run up to k-steps, dump to local files
- Dump the output of each step using a different backend(opencl or cpu)
- Check at which specific step the result went wrong, compare the generated code, use the dumped data to generate a minimum reproducible example.
- Reduce the schedule of the errored code, to a minimum one to see if there is anything wrong.

**Comment by headupinclouds** [comment]
I reported this issue in the post you mentioned [here](https://discuss.tvm.ai/t/vulkan-deploy-issue-with-latest-code/604/7).  

> Get an end to end example that produces the wrong result

I can post a small sample to reproduce this for review and discussion in a github repo.

> Use debug runtime to run up to k-steps, dump to local files ...

I was looking for mechanisms to support this from C++ (asked [here](https://discuss.tvm.ai/t/layer-serialization-from-c/1406)).   Can you share any pointers or examples that might help illustrate how to perform such logging from C++?

**Comment by tqchen** [comment]
You can setup an  RPC server in the mobile and use https://docs.tvm.ai/api/python/graph_runtime.html#tvm.contrib.graph_runtime.GraphModule.debug_get_output to get k-th output. Then you can just log the output using numpy's typical serialization mechanism

**Comment by headupinclouds** [comment]
Thanks.  That pointed me in the right direction.  I'm able to use the same mechanism from C++ with `tvm.graph_runtime_debug.create` and the `get_output_by_layer` packaged function calls.

https://gist.github.com/headupinclouds/d87df2dcc9603589b3b9b9cb00f26011#file-tvm_deploy_gpu_sample-cpp-L521

If I use `opt_level=0` the example does return the correct result on the same Android device.  It is only when `opt_level > 0` where things break down.

```
with nnvm.compiler.build_config(opt_level=0):
    graph, lib, params = nnvm.compiler.build(sym, target, shape_dict, params=params, target_host=target_host)
```

The same code with `opt_level=3` (any >= 1) works on Android with `llvm` (cpu) and on the Ubuntu host for all tested back-ends: `llvm` (cpu), `OpenGL`, `Vulkan`, `OpenCL`, and `gpu` (cuda).

I'll see if I can understand the root cause.

**Comment by headupinclouds** [comment]
I've posted an initial example to help reproduce the issue here:

https://github.com/headupinclouds/tvm_cpp_test

In addition to the observations noted above (the same example works in every tested configuration except Android w/ Vulkan), I've noticed the error:

* is reproducible across runs (consistent but wrong)
* occurs in the first actual `nnvm` fused layer (`fuse_conv2d_broadcast_add_relu` below)
* is eliminated by using `opt_level=0`

`head` of compiled graph `from_mxnet.json` w/ `opt_level=3` where output is incorrect (fusion)
```
{
  "nodes": [
    {
      "op": "null", 
      "name": "data", 
      "inputs": []
    }, 
    {
      "op": "null", 
      "name": "resnetv10_conv0_weight_sc", 
      "inputs": []
    }, 
    {
      "op": "null", 
      "name": "batch_norm0_add_beta_expand", 
      "inputs": []
    }, 
    {
      "op": "tvm_op", 
      "name": "relu0", 
      "attrs": {
        "flatten_data": "0", 
        "func_name": "fuse_conv2d_broadcast_add_relu", 
        "num_inputs": "3", 
        "num_outputs": "1"
      }, 
      "inputs": [[0, 0, 0], [1, 0, 0], [2, 0, 0]]
    }, 
```

`head` of compiled graph `from_mxnet.json` w/ `opt_level=0` where output is okay (no fusion)
```
{
  "nodes": [
    {
      "op": "null", 
      "name": "data", 
      "inputs": []
    }, 
    {
      "op": "null", 
      "name": "resnetv10_conv0_weight", 
      "inputs": []
    }, 
    {
      "op": "tvm_op", 
      "name": "conv2d0", 
      "attrs": {
        "flatten_data": "0", 
        "func_name": "fuse_conv2d", 
        "num_inputs": "2", 
        "num_outputs": "1"
      }, 
      "inputs": [[0, 0, 0], [1, 0, 0]]
    }, 
    {
      "op": "null", 
      "name": "resnetv10_batchnorm0_running_var", 
      "inputs": []
    }, 
    {
      "op": "tvm_op", 
      "name": "batch_norm0_add_eps", 
      "attrs": {
        "flatten_data": "1", 
        "func_name": "fuse___add_scalar__", 
        "num_inputs": "1", 
        "num_outputs": "1"
      }, 
      "inputs": [[3, 0, 1]]
    }, 
```

**Comment by eqy** [comment]
Thanks for your work on this! Were you able to verify that it was the fusion pass in particular that produces the error (e.g., by disabling/enabling other passes to check https://discuss.tvm.ai/t/different-output-values-when-setting-opt-level-3-in-nnvm-compiler-build-config/1392).

**Comment by headupinclouds** [comment]
> Were you able to verify that it was the fusion pass in particular that produces the error `opt_level=2` triggers an error.

I don't know enough to say that specifically.

Here is the output for opt_level=0,1,2,3.

### opt_level=3

```
The maximum position in output vector is: 669
Expected 282 but got: 669
```

### opt_level=2

```
error: [21:36:32] /dl/mxnet/3rdparty/tvm/apps/howto_deploy/../../src/runtime/module_util.cc:53: Check failed: ret == 0 (-1 vs. 0) [21:36:32] /dl/mxnet/3rdparty/tvm/apps/howto_deploy/../../src/runtime/vulkan/vulkan_module.cc:328: Check failed: __e == VK_SUCCESS Vulan Error, code=-3: VK_ERROR_INITIALIZATION_FAILED
terminating with uncaught exception of type dmlc::Error: [21:36:32] /dl/mxnet/3rdparty/tvm/apps/howto_deploy/../../src/runtime/vulkan/vulkan_device_api.cc:345: Check failed: __e == VK_SUCCESS Vulan Error, code=-3: VK_ERROR_INITIALIZATION_FAILED
Aborted 
```

### opt_level=1

```
The maximum position in output vector is: 574
Expected 282 but got: 574
```

### opt_level=0

```
The maximum position in output vector is: 282
```

Please let me know if there are other tests that could be useful.

Is it possible to debug this at the level of individual Vulkan commands somehow?

**Comment by eqy** [comment]
Maybe a way to simplify the debugging is to see if the behavior changes for smaller pieces of the graph. As it is, it may be difficult to trace exactly the error is when doing a full end-to-end inference. Can you try checking what happens, say, when only running the first layer---adding layers until the problem resurfaces?

**Comment by headupinclouds** [comment]
Thanks for the feedback.  In this case it actually does seem to occur directly in the first real layer of the fused/optimized network `fuse_conv2d_broadcast_add_relu` in the post [above](https://github.com/dmlc/tvm/issues/2355#issuecomment-451781350).

**Comment by eqy** [comment]
Thanks for this; it is useful to know that the problem can be triggered with a single fused conv2d layer.
I think the next step is to check if the problem persists when using a default schedule for the operator.
For example, we could manually define a direct 2d convolution with minimal scheduling (e.g., just bind threads to a single dimension) while inlining the next operation (to simulate fusion).

**Comment by headupinclouds** [comment]
> I think the next step is to check if the problem persists when using a default schedule for the operator.

Okay, this sounds good, although I'll need help with the details (unless you have an example close to what you are suggesting already).  It sounds like this would be a small unit test entirely in TVM (no NNVM or end-to-end CNN) that runs something like conv2d + relu on a simple test input tensor `ones(64,64,3)`?  We would make a small python script, then `build` for Android + Vulkan and run the test as a `C++` application on the device.  Right?

Maybe the following TVM tutorial is a reasonable starting point?

https://docs.tvm.ai/tutorials/topi/intro_topi.html#fusing-convolutions

Are there other generic unit tests we should be running (in addition) to help test the Vulkan back-end on Android?  Maybe these https://github.com/dmlc/tvm/tree/master/tests/cpp?

**Comment by eqy** [comment]
Yes, your understanding is correct. That TVM tutorial should be almost exactly what we need.
If you have a C++ function deploy flow on device, we can use that, but otherwise it would be simpler to just use the RPC server so that we can make changes quickly without rebuilding anything manually for the device.

I am not sure we need to use the packed function test, because more basic functions (e.g., vector-add, or even un-fused conv2d would tell us that the runtime is working at a more primitive level).

Basically the purpose of this is to see if the problem lies within the schedule somewhere, or occurs any time we invoke fusion. I hope it is not a corner case where a specific schedule + fusion are needed to trigger it.

**Comment by headupinclouds** [comment]
I made a smaller test project [tvm_intro_topi](https://github.com/headupinclouds/tvm_intro_topi/blob/69a709e48ddb53ad7856c4c18bb99ea9f20d58c5/intro_topi.py#L101-L104) containing just the conv2d + relu chain along with a C++ test executable.

```
data = tvm.placeholder((1, 3, 224, 224))
kernel = tvm.placeholder((10, 3, 5, 5))

with tvm.target.create("cuda"):
    conv = topi.nn.conv2d(data, kernel, strides=1, padding=2, dilation=1)
    out = topi.nn.relu(conv)
    sconv = topi.generic.nn.schedule_conv2d_nchw(out)
    print(tvm.lower(sconv, [data, kernel], simple_mode=True))
```

This is based on the original TVM TOPI tutorial [intro_topi.py](https://github.com/dmlc/tvm/blob/45456b1495048883fa3b3f2095c6747b8188db85/tutorials/topi/intro_topi.py#L116-L123) we discussed above.

Both the C++ and Vulkan versions run correctly on the same Android test device.

> we would manually schedule the fused convolution (as minimally as possible)

What modifications do you recommend to the `topi_intro.py` script to achieve this?

Thanks for the help.

**Comment by headupinclouds** [comment]
The from_mxnet.py example is using [resnet18_v](https://github.com/headupinclouds/tvm_cpp_test/blob/9c41ead301545120975d55c3806689d7e610bb71/from_mxnet.py#L78)

```
block = get_model('resnet18_v1', pretrained=True)
```

Since we seem to see issues in the first block, we can probably replicate that from the gluon-cv definition  [here](https://github.com/dmlc/gluon-cv/blob/5e7a3a2f86d0fe4c8ad394c03618c7abca07f608/gluoncv/model_zoo/resnet.py#L392-L395).  I guess we want some (or all) of this part:

```
                self.features.add(nn.Conv2D(channels[0], 7, 2, 3, use_bias=False))
                self.features.add(norm_layer(**({} if norm_kwargs is None else norm_kwargs)))
                self.features.add(nn.Activation('relu'))
                self.features.add(nn.MaxPool2D(3, 2, 1))
```

The details of how to replicate the NNVM fusion (in particular, replicating `OpFusion`) are not currently clear to me.  I'll try read through that code in more detail.

**Comment by eqy** [comment]
I think a simple thing to try would be to just tack on a few elementwise operations (e.g., add, subtract, multiply) following the convolution, and then trying `compute_inline`.
https://docs.tvm.ai/tutorials/language/schedule_primitives.html#compute-inline

**Comment by tqchen** [comment]
Do we have any followup on this @headupinclouds @eqy ?

**Comment by headupinclouds** [comment]
I added explicit batch-norm affine arithmetic and experimented with `compute_inline` in the test repo I posted, but it didn't seem to trigger an error.  I'll push the updated changes.   I guess we can try to translate the head of the resent18 example into the standalone test project to help isolate things.

**Comment by nscotto** [comment]
Hi. I am getting mismatched result when running [apps/android_rpc/tests/android_rpc_test.py](The example just runs a sum on the device, so there are no conv2d at all) for vulkan, the example just runs a sum on the device, so there are no conv2d at all did it already happen to you?

**Comment by tqchen** [comment]
NOTE: the vulkan runtime has been rewritten, would be great to confirm again if the result is correct under the new runtime.

**Comment by tqchen** [comment]
close for now due to inactive status, we will open new thread to track problems if there are more followups
